plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/FloatDelay.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE LambdaCase #-}
+
+module UntypedPlutusCore.Transform.FloatDelay (floatDelay) where
+
+import PlutusCore qualified as PLC
+import UntypedPlutusCore.Core
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Writer.CPS
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+{- | The Float Delay optimization floats `Delay` from arguments into function bodies,
+if possible. It turns @(\n -> ...Force n...Force n...) (Delay arg)@ into
+@(\n -> ...Force (Delay n)...Force (Delay n)...) arg@.
+
+The above transformation is performed if:
+
+    * All occurrences of @arg@ are under @Force@.
+
+    * @arg@ is essentially work-free.
+
+This achieves a similar effect to Plutonomy's "Split Delay" optimization. The difference
+is that Split Delay simply splits the @Delay@ argument into multiple arguments, turning the
+above example into @(\m -> (\n -> ...Force n...Force n...) (Delay m)) arg@, and then relies
+on other optimizations to simplify it further. Specifically, once the inliner inlines
+@Delay m@, it will be identical to the result of Float Delay.
+
+The advantages of Float Delay are:
+
+    * It is very simple to implement.
+
+    * It doesn't rely on the inliner. In this example, Split Delay relies on the inliner to
+      inline @Delay m@, but there's no guarantee that the inliner will do so, because inlining
+      it may increase the program size.
```

Comment:
> If the inliner doesn't want to inline some thing but FloatDelay does, then one of them is wrong.

I don't agree with this. The inliner has no idea that, after inlining `Delay`, it will be cancelled by `ForceDelayCancel`. On its face, inlining `Delay` increases code size.

We can make the inliner aware of the `Force`s and the `ForceDelayCancel`, but then the different optimization passes are _more_ intertwined, not less, IMO.

Reply:
It's context-sensitive inlining, sure, but in that respect it's no different from inlining saturated functions or inlining constructor applications into case contexts. You need to know a bit about the reduction behaviour, sure, but that's not so crazy.

Reply:
Basically an optimization pass may know better than the inliner, and it can be awkward to duplicate that knowledge in the inliner. Take a look at https://github.com/input-output-hk/plutus/pull/5308. I believe it is an even stronger case of "don't rely on the inliner".

Reply:
I still disagree but good to write it down :)